### PR TITLE
Changes 2025

### PR DESCRIPTION
**Pull request by afoix**
Summary of the changes that the PR contains:

- [x] Correct typos
- [x] Add the answer to the bonus questions in case they want to use the all-grid datasets or all-dots dataset 
- [x] Remove old text from 2023
- [x] Add the learning goals in each Checkpoint as requested in #7

### CODE REVIEW

File: solution.py [hunk 1]
```diff
@@ -642,24 +734,23 @@ def cm_analysis(y_true, y_pred, title, figsize=(10,10)):
...
+# and local corruption used both some true and some false features about 7s.
+# Ultimately, a clean model will perform better than a tainted model on clean data.
 
 # %% [markdown]
 # <div class="alert alert-success"><h3>
 #     Checkpoint 3</h3>
 #
 # Post to the course chat when you have reached Checkpoint 3, and will will discuss our results and reasoning about why they might have happened.
+# <h4> Learning goals of part 3</h4>
+# In this thrid part of the exercise we've learned:
```
Comment (cmalinmayor) [inline solution.py]:
```suggestion
# In this third part of the exercise we've learned:
```

File: solution.py [hunk 2]
```diff
@@ -839,19 +984,19 @@ def visualize_integrated_gradients(test_input, model, plot_title):
...
-#
-# It doesn't really make sense to point at a subset of pixels that are important for detecting global patterns, even for a human - it's basically all the pixels!
-
 # %% [markdown]
 # <div class="alert alert-block alert-success"><h3>
 #     Checkpoint 4</h3>
 #     <ol>
-#         Congrats on finishing the integrated gradients task! Let us know on the course chat that you reached checkpoint 4, and feel free to look at other interpretability methods in the Captum library if you're interested.
+#         Congrats on finishing the integrated gradients task! Let us know on the course chat that you reached checkpoint 4, a
+# nd feel free to look at other interpretability methods in the Captum library if you're interested.
```
Comment (cmalinmayor) [inline solution.py]:
```suggestion
#         Congrats on finishing the integrated gradients task! Let us know on the course chat that you reached checkpoint 4,
# and feel free to look at other interpretability methods in the Captum library if you're interested.
```

File: solution.py [hunk 3]
```diff
@@ -92,25 +102,25 @@
 # First we will add a white pixel in the bottom right of all images of 7's, and visualize the results. This is an example of a local change to the images, where only a small portion of the image is corrupted.
 
 # %%
-# Add a white pixel in the bottom right of all images of 7's
-tainted_train_dataset.data[train_dataset.targets==7, 25, 25] = 255
-tainted_test_dataset.data[test_dataset.targets==7, 25, 25] = 255
+# Add for white white pixels in the bottom right of all images of 7's
```
Comment (cmalinmayor) [inline solution.py]:
```suggestion
# Add four white white pixels in the bottom right of all images of 7's
```

Comment (cmalinmayor) [reply solution.py]: And actually it looks like it's 9 pixels?

File: solution.py [hunk 4]
```diff
@@ -584,19 +687,15 @@ def cm_analysis(y_true, y_pred, title, figsize=(10,10)):
...
 # **3.1 Answer:**
 #
-# The clean model on the clean dataset predicted 5s least accurately, with some confusion with 6s and 3s. These are likely confused by the model as handwritten 5s may look like 6s (almost closed bottom part) or 3s (presence of 3 horizontal segments).
-
-# %% [markdown] tags=["solution"]
-# **3.1 Answer from 2023 Students**
-#
-# 5 is the least accurately predicted digit. It is most confused with 6 or 3.
-# Handwriting creates fives that look like sixes or threes.
+# The clean model on the clean dataset predicted 5s least accurately, with some confusion with 6s and 3s.
```
Comment (cmalinmayor) [inline solution.py:697]: My clean model confused 4/9 and 7/9 most - presumably this is due to randomness, so perhaps we should edit the solution to say that their model might be different.

File: solution.py [hunk 5]
```diff
@@ -642,24 +734,23 @@ def cm_analysis(y_true, y_pred, title, figsize=(10,10)):
...
-# **3.4 Answer from 2023 Students:**
-#
-# Clean 7s vs clean 4s: 4 WINS! (aka is worse)
-#
-# Global corruptions are more detrimental when testing on the clean data. This is because the training images are *more* different from each other.
-#
-# Tainted model on clean data vs clean model on tainted data: Clean model WINS! (is better on tainted data than tainted model on clean data)
-#
-# The clean model still has useful signal to work with in the tainted data. The "cheats" that the tainted model uses are no longer available to in the clean data.
+# The tainted model performed poorly on clean 7s and extremely poorly on clean 4s.
```
Comment (cmalinmayor) [inline solution.py:738]: My tainted model also got 0% on both clean 4s and 7s 🙃 so again mentioning it might be a bit random

File: solution.py [hunk 6]
```diff
@@ -863,6 +1008,88 @@ def visualize_integrated_gradients(test_input, model, plot_title):
...
+)
+visualize_integrated_gradients(
+    test_dataset[1], model_allgrid, "All-grid Model on clean data"
+)
+
+
+# %%  [markdown] tags=["solution"]
+# **Bonus question answer:**
+# Its all classified in the same way, with the grid pattern being ignored.
+# All the inputs are lying in the same category, numbr 1 in both cases when all-grid model in all-grid datat and in clean data all classified as 1.
```
Comment (cmalinmayor) [inline solution.py]:
```suggestion
# All the inputs are lying in the same category, number 1 in both cases when all-grid model in all-grid data and in clean data all classified as 1.
```

File: solution.py [hunk 7]
```diff
@@ -1125,13 +1369,9 @@ def visualize_denoising(model, dataset, index):
...
 # **5.3 Answer:**
 #
-# If a denoising model is trained on data which does not appear in the data it is ultimatly used on, that new content will end up likely changed. A real worl example could be that of training a model on lots of non-dividing cells images, and use the model on new data which happens to contain some dividing cells. This could lead to the information being "denoised" away.
-
-# %% [markdown] tags=["solution"]
-# **5.3 Answer from 2023**
-#
-# - Run on any out of distribution data
-# - Especially tricky if the data appears to be in distribution but has rare events. E.g. if the denoiser was trained on lots of cells that were never dividing and then was run on similar image with dividing cells, it might remove the dividing cell and replace with a single cell.
+# If a denoising model is trained on data which does not appear in the data it is ultimatly used on,
```
Comment (cmalinmayor) [inline solution.py]:
```suggestion
# If a denoising model is trained on data which does not appear in the data it is ultimately used on,
```

**Review by cmalinmayor — COMMENTED** [review]
Amazing! I love this exercise. I ran it on the VM and it worked smoothly. I left a few comments about typos and randomness in solutions. My only other note is that training the denoising model takes about 1.5 (only mnist) or 2.5 (mnist and fashion) minutes per epoch, and we are training 5 epochs -> 10ish minutes is kind of long. The training loss seems to saturate early - perhaps we can train fewer epochs? I tried training only two epochs for the original digit denoiser and it definitely performed a little worse, so maybe 3 is the sweet spot? If we need to train longer, we should at least add a note that "this will take 5-10 minutes."

File: solution.py (hunk 3 above)
Comment (afoix) [reply solution.py]:
```
tainted_train_dataset.data[train_dataset.targets == 7, 24:27, 24:27] = 255
```

You are right they are 9 pixels!! Sorry about that! I am updating that too!

File: solution.py (hunk 4 above)
Comment (afoix) [reply solution.py:697]: Also added -> consider randomness!

**Comment by afoix** [comment]
@cmalinmayor thank you for your feedback! It was very useful 😄  I've corrected all the typos, added the sentences indicating that there is some randomness involved and lastly, I modify the training so instead of training for 5 epochs, we train for 3 epochs as you suggested! Let me know if it looks acceptable and I will merge the PR 🥳

**Comment by cmalinmayor** [comment]
Looks good to me! 🙂 thanks!

**Review by cmalinmayor — APPROVED** [review]
